From b99c6a540a00d99b1cf71b468066a1a11a3f1751 Mon Sep 17 00:00:00 2001
From: Florian Westphal <fwestpha@redhat.com>
Date: Thu, 24 Jul 2014 12:29:03 -0400
Subject: [net] netfilter: connlimit: improve packet-to-closed-connection logic

Message-id: <1406204949-17382-3-git-send-email-fwestpha@redhat.com>
Patchwork-id: 86623
O-Subject: [PATCH 2/8] [RHEL7.1 net] netfilter: connlimit: improve packet-to-closed-connection logic
Bugzilla: 1052282
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

BZ: 1052282

commit d9ec4f1ee280e5f8732e3c40ca672419b2532600
Author: Florian Westphal <fw@strlen.de>
Date:   Fri Mar 7 14:37:10 2014 +0100

    netfilter: connlimit: improve packet-to-closed-connection logic

    Instead of freeing the entry from our list and then adding
    it back again in the 'packet to closing connection' case just keep the
    matching entry around.  Also drop the found_ct != NULL test as
    nf_ct_tuplehash_to_ctrack is just container_of().

    Reviewed-by: Jesper Dangaard Brouer <brouer@redhat.com>
    Signed-off-by: Florian Westphal <fw@strlen.de>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/xt_connlimit.c b/net/netfilter/xt_connlimit.c
index 6988818..d4c6db1 100644
--- a/net/netfilter/xt_connlimit.c
+++ b/net/netfilter/xt_connlimit.c
@@ -112,29 +112,22 @@ static int count_hlist(struct net *net,
  hlist_for_each_entry_safe(conn, n, head, node) {
   found    = nf_conntrack_find_get(net, NF_CT_DEFAULT_ZONE,
        &conn->tuple);
-  found_ct = NULL;
+  if (found == NULL) {
+   hlist_del(&conn->node);
+   kfree(conn);
+   continue;
+  }
 
-  if (found != NULL)
-   found_ct = nf_ct_tuplehash_to_ctrack(found);
+  found_ct = nf_ct_tuplehash_to_ctrack(found);
 
-  if (found_ct != NULL &&
-      nf_ct_tuple_equal(&conn->tuple, tuple) &&
-      !already_closed(found_ct))
+  if (nf_ct_tuple_equal(&conn->tuple, tuple)) {
    /*
     * Just to be sure we have it only once in the list.
     * We should not see tuples twice unless someone hooks
     * this into a table without "-p tcp --syn".
     */
    addit = false;
-
-  if (found == NULL) {
-   /* this one is gone */
-   hlist_del(&conn->node);
-   kfree(conn);
-   continue;
-  }
-
-  if (already_closed(found_ct)) {
+  } else if (already_closed(found_ct)) {
    /*
     * we do not care about connections which are
     * closed already -> ditch it
-- 
1.7.1